subsys/shell/shell_mqtt.c

```diff
@@ -0,0 +1,659 @@
+			(void)ring_buf_put(&sh_mqtt->rx_rb, "\r\n", sizeof("\r\n"));
+		} else {
+			LOG_ERR("RB free space: %d bytes, incoming packet: %d bytes", rb_free_space,
+				pub->message.payload.len);
+		}
+
+		sh_mqtt->shell_handler(SHELL_TRANSPORT_EVT_RX_RDY, sh_mqtt->shell_context);
+		break;
+	}
+
+	case MQTT_EVT_PUBACK:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT PUBACK error %d", evt->result);
+			break;
+		}
+		LOG_DBG("PUBACK packet id: %u", evt->param.puback.message_id);
+		break;
+
+	case MQTT_EVT_PINGRESP:
+		LOG_DBG("PINGRESP packet");
+		break;
+
+	default:
+		LOG_DBG("MQTT event received %d", evt->type);
+		break;
+	}
+}
+
+static void broker_init(struct shell_mqtt *ctx)
+{
+#ifdef CONFIG_NET_IPV6
+	struct sockaddr_in6 *broker6 = (struct sockaddr_in6 *)&ctx->broker;
+
+	broker6->sin6_family = AF_INET6;
+	broker6->sin6_port = htons(SERVER_PORT);
+	(void)zsock_inet_pton(AF_INET6, SERVER_ADDR, &broker6->sin6_addr);
+
+#ifdef CONFIG_SOCKS
+	struct sockaddr_in6 *proxy6 = (struct sockaddr_in6 *)&ctx->socks5_proxy;
+
+	proxy6->sin6_family = AF_INET6;
+	proxy6->sin6_port = htons(SOCKS5_PROXY_PORT);
+	(void)zsock_inet_pton(AF_INET6, SOCKS5_PROXY_ADDR, &proxy6->sin6_addr);
+#endif
+#else /* IPV4 */
+	struct sockaddr_in *broker4 = (struct sockaddr_in *)&ctx->broker;
+
+	broker4->sin_family = AF_INET;
+	broker4->sin_port = htons(SERVER_PORT);
+	(void)zsock_inet_pton(AF_INET, SERVER_ADDR, &broker4->sin_addr);
+#ifdef CONFIG_SOCKS
+	struct sockaddr_in *proxy4 = (struct sockaddr_in *)&ctx->socks5_proxy;
+
+	proxy4->sin_family = AF_INET;
+	proxy4->sin_port = htons(SOCKS5_PROXY_PORT);
+	(void)zsock_inet_pton(AF_INET, SOCKS5_PROXY_ADDR, &proxy4->sin_addr);
+#endif
+#endif
+}
+
+static struct mqtt_utf8 password = { .utf8 = MQTT_PASSWORD, .size = sizeof(MQTT_PASSWORD) };
+
+static struct mqtt_utf8 username = { .utf8 = MQTT_USERNAME, .size = sizeof(MQTT_USERNAME) };
+
+static void client_init(struct shell_mqtt *ctx)
+{
+	mqtt_client_init(&ctx->client_ctx);
+
+	/* MQTT client configuration */
+	ctx->client_ctx.broker = &ctx->broker;
+	ctx->client_ctx.evt_cb = mqtt_evt_handler;
+	ctx->client_ctx.client_id.utf8 = (uint8_t *)sh_mqtt->device_id;
+	ctx->client_ctx.client_id.size = strlen(sh_mqtt->device_id);
+	ctx->client_ctx.password = &password;
+	ctx->client_ctx.user_name = &username;
+	ctx->client_ctx.protocol_version = MQTT_VERSION_3_1_1;
+
+	/* MQTT buffers configuration */
+	ctx->client_ctx.rx_buf = ctx->buf.rx;
+	ctx->client_ctx.rx_buf_size = sizeof(ctx->buf.rx);
+	ctx->client_ctx.tx_buf = ctx->buf.tx;
+	ctx->client_ctx.tx_buf_size = sizeof(ctx->buf.tx);
+
+	/* MQTT transport configuration */
+#ifdef CONFIG_MQTT_LIB_WEBSOCKET
+	ctx->client_ctx.transport.type = MQTT_TRANSPORT_NON_SECURE_WEBSOCKET;
+#else
+	ctx->client_ctx.transport.type = MQTT_TRANSPORT_NON_SECURE;
+#endif
+
+#ifdef CONFIG_MQTT_LIB_WEBSOCKET
+	ctx->client_ctx.transport.websocket.config.host = SERVER_ADDR;
+	ctx->client_ctx.transport.websocket.config.url = "/shell_mqtt";
```

Comment:
Did you test this wih any server, does it work? I'm wondering, because MQTT spec says that for websocket transport the URI should be plain `/mqtt`, but maybe it's not a hard requirement.

Reply:
No, I only tested the normal IPV4 connection, the other are referenced from: https://github.com/zephyrproject-rtos/zephyr/blob/main/samples/net/cloud/google_iot_mqtt/src/protocol.c